x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/ingest/InferenceProcessor.java

```diff
@@ -53,26 +140,122 @@ public String getType() {
+                        ingestService.getScriptService());
+                    count += pipeline.getProcessors().stream().filter(processor -> processor instanceof InferenceProcessor).count();
+                } catch (Exception ex) {
+                    logger.warn(new ParameterizedMessage("failure parsing pipeline config [{}]", configuration.getId()), ex);
+                }
+            }
+            currentInferenceProcessors = count;
+        }
+
+        // Used for testing
+        int numInferenceProcessors() {
+            return currentInferenceProcessors;
         }
 
         @Override
-        public Processor create(Map<String, Processor.Factory> processorFactories, String tag, Map<String, Object> config)
+        public InferenceProcessor create(Map<String, Processor.Factory> processorFactories, String tag, Map<String, Object> config)
             throws Exception {
+
+            if (this.maxIngestProcessors <= currentInferenceProcessors) {
+                throw new ElasticsearchStatusException("Max number of inference processors reached, total inference processors [{}]. " +
+                    "Adjust the setting [{}]: [{}] if a greater number is desired.",
+                    RestStatus.CONFLICT,
+                    currentInferenceProcessors,
+                    MAX_INFERENCE_PROCESSORS.getKey(),
+                    maxIngestProcessors);
+            }
+
             String modelId = ConfigurationUtils.readStringProperty(TYPE, tag, config, MODEL_ID);
-            return new InferenceProcessor(client, tag, modelId);
+            String targetField = ConfigurationUtils.readStringProperty(TYPE, tag, config, TARGET_FIELD);
+            Map<String, String> fieldMapping = ConfigurationUtils.readOptionalMap(TYPE, tag, config, FIELD_MAPPINGS);
+            InferenceConfig inferenceConfig = inferenceConfigFromMap(ConfigurationUtils.readMap(TYPE, tag, config, INFERENCE_CONFIG));
+            String modelInfoField = ConfigurationUtils.readStringProperty(TYPE, tag, config, MODEL_INFO_FIELD, "_model_info");
+            if (modelInfoField != null && tag != null) {
+                modelInfoField += "." + tag;
+            }
+            return new InferenceProcessor(client, tag, targetField, modelId, inferenceConfig, fieldMapping, modelInfoField);
         }
 
-        @Override
-        public void accept(ClusterState clusterState) {
+        // Package private for testing
+        void setMaxIngestProcessors(int maxIngestProcessors) {
+            logger.trace("updating setting maxIngestProcessors from [{}] to [{}]", this.maxIngestProcessors, maxIngestProcessors);
+            this.maxIngestProcessors = maxIngestProcessors;
+        }
+
+        InferenceConfig inferenceConfigFromMap(Map<String, Object> inferenceConfig) throws IOException {
+            ExceptionsHelper.requireNonNull(inferenceConfig, INFERENCE_CONFIG);
+
+            if (inferenceConfig.keySet().size() != 1) {
```

Comment:
Is it equivalent to `inferenceConfig.size()`? It should be, right?

Reply:
definitely, I can simplify.